### PR TITLE
follow

### PR DESCRIPTION
**Pull request by tranduytung99**

### CODE REVIEW

File: app/controllers/relationships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class RelationshipsController < ApplicationController
+  # before_action :logged_in_user
```
Comment (huongnguyenmta) [inline app/controllers/relationships_controller.rb]: remove comment

File: app/controllers/relationships_controller.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  def create
+    user = User.find(params[:followed_id])
```
Comment (huongnguyenmta) [inline app/controllers/relationships_controller.rb]: using `find_by`

File: app/controllers/relationships_controller.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    current_user.follow(user)
```
Comment (huongnguyenmta) [inline app/controllers/relationships_controller.rb]: remove ()

File: app/controllers/relationships_controller.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    redirect_to user
+  end
+
+  def destroy
+    user = Relationship.find(params[:id]).followed
+    current_user.unfollow(user)
```
Comment (huongnguyenmta) [inline app/controllers/relationships_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -51,6 +51,20 @@ def destroy
     end
   end
 
+  def following
+    @title = "Following"
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: I18n

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -51,6 +51,20 @@ def destroy
...
+    @user  = User.find(params[:id])
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: using `find_by`

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -51,6 +51,20 @@ def destroy
...
+    @users = @user.following.paginate(page: params[:page])
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -51,6 +51,20 @@ def destroy
...
+    render 'show_follow'
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: using double quote

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -51,6 +51,20 @@ def destroy
...
+  end
+
+  def followers
+    @title = "Followers"
+    @user  = User.find(params[:id])
+    @users = @user.followers.paginate(page: params[:page])
+    render 'show_follow'
+  end
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: C fix đoạn code này theo comment ở trên nhé

File: app/models/user.rb [hunk 10]
```diff
@@ -35,6 +42,18 @@ def authenticated? attribute, token
     BCrypt::Password.new(digest).is_password?(token)
   end
 
+  def follow(other_user)
```
Comment (huongnguyenmta) [inline app/models/user.rb]: remove ()

File: app/models/user.rb [hunk 11]
```diff
@@ -35,6 +42,18 @@ def authenticated? attribute, token
...
+    following << other_user
+  end
+  
+  def unfollow(other_user)
```
Comment (huongnguyenmta) [inline app/models/user.rb]: remove ()

File: app/models/user.rb [hunk 12]
```diff
@@ -35,6 +42,18 @@ def authenticated? attribute, token
...
+    following.delete(other_user)
```
Comment (huongnguyenmta) [inline app/models/user.rb]: remove ()

File: app/models/user.rb [hunk 13]
```diff
@@ -35,6 +42,18 @@ def authenticated? attribute, token
...
+  end
+
+  def following?(other_user)
```
Comment (huongnguyenmta) [inline app/models/user.rb]: remove ()

File: app/models/user.rb [hunk 14]
```diff
@@ -35,6 +42,18 @@ def authenticated? attribute, token
...
+    following.include?(other_user)
```
Comment (huongnguyenmta) [inline app/models/user.rb]: remove ()

File: app/views/shared/_stats.html.erb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+<% @user ||= current_user %>
+<div class="stats">
+  <a href="<%= following_user_path(@user) %>">
+    <strong id="following" class="stat">
+      <%= @user.following.count %>
+    </strong>
+    following
```
Comment (huongnguyenmta) [inline app/views/shared/_stats.html.erb]: I18n

File: app/views/shared/_stats.html.erb [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+  </a>
+  <a href="<%= followers_user_path(@user) %>">
```
Comment (huongnguyenmta) [inline app/views/shared/_stats.html.erb]: remove ()

File: app/views/shared/_stats.html.erb [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+    <strong id="followers" class="stat">
+      <%= @user.followers.count %>
+    </strong>
+    followers
+  </a>
+</div>
```
Comment (huongnguyenmta) [inline app/views/shared/_stats.html.erb]: add line space in end this file

File: app/views/static_pages/home.html.erb [hunk 18]
```diff
@@ -5,6 +5,9 @@
       <section class="user_info">
         <%= render "shared/user_info" %>
       </section>
+      <section class="stats">
+        <%= render 'shared/stats' %>
```
Comment (huongnguyenmta) [inline app/views/static_pages/home.html.erb]: using double quote

File: app/views/users/_follow.html.erb [hunk 19]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.build) do |f| %>
+  <div><%= hidden_field_tag :followed_id, @user.id %></div>
+  <%= f.submit "Follow", class: "btn btn-primary" %>
+<% end %>
```
Comment (huongnguyenmta) [inline app/views/users/_follow.html.erb]: add line space in end this file

File: app/views/users/_follow.html.erb [hunk 20]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.build) do |f| %>
+  <div><%= hidden_field_tag :followed_id, @user.id %></div>
+  <%= f.submit "Follow", class: "btn btn-primary" %>
```
Comment (huongnguyenmta) [inline app/views/users/_follow.html.erb]: I18n

File: app/views/users/_follow_form.html.erb [hunk 21]
```diff
@@ -0,0 +1,9 @@
+<% unless current_user?(@user) %>
+  <div id="follow_form">
+  <% if current_user.following?(@user) %>
```
Comment (huongnguyenmta) [inline app/views/users/_follow_form.html.erb]: remove ()

File: app/views/users/_follow_form.html.erb [hunk 22]
```diff
@@ -0,0 +1,9 @@
...
+    <%= render 'unfollow' %>
```
Comment (huongnguyenmta) [inline app/views/users/_follow_form.html.erb]: using double quote

File: app/views/users/_follow_form.html.erb [hunk 23]
```diff
@@ -0,0 +1,9 @@
...
+  <% else %>
+    <%= render 'follow' %>
```
Comment (huongnguyenmta) [inline app/views/users/_follow_form.html.erb]: using double quote

File: app/views/users/_follow_form.html.erb [hunk 24]
```diff
@@ -0,0 +1,9 @@
...
+  <% end %>
+  </div>
+<% end %>
```
Comment (huongnguyenmta) [inline app/views/users/_follow_form.html.erb]: add line space in end this file

File: app/views/users/_unfollow.html.erb [hunk 25]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.find_by(followed_id: @user.id),
```
Comment (huongnguyenmta) [inline app/views/users/_unfollow.html.erb]: remove ()

File: app/views/users/_unfollow.html.erb [hunk 26]
```diff
@@ -0,0 +1,4 @@
...
+             html: { method: :delete }) do |f| %>
```
Comment (huongnguyenmta) [inline app/views/users/_unfollow.html.erb]: indent

File: app/views/users/_unfollow.html.erb [hunk 27]
```diff
@@ -0,0 +1,4 @@
...
+  <%= f.submit "Unfollow", class: "btn" %>
+<% end %>
```
Comment (huongnguyenmta) [inline app/views/users/_unfollow.html.erb]: add line space in end this file

File: app/controllers/users_controller.rb [hunk 28]
```diff
@@ -51,6 +51,18 @@ def destroy
     end
   end
 
+  def following
+    @title = t ".title"
+    @users = @user.following.paginate(page: params[:page])
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 29]
```diff
@@ -51,6 +51,18 @@ def destroy
...
+    render "show_follow"
+  end
+
+  def followers
+    @title = "Followers"
+    @users = @user.followers.paginate(page: params[:page])
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: remove ()

File: app/models/user.rb [hunk 30]
```diff
@@ -1,5 +1,12 @@
 class User < ApplicationRecord
   has_many :microposts, dependent: :destroy
+  has_many :active_relationships, class_name: "Relationship",
+    foreign_key: "follower_id", dependent: :destroy
+  has_many :passive_relationships, class_name: "Relationship",
+    foreign_key: "followed_id", dependent:   :destroy
```
Comment (huongnguyenmta) [inline app/models/user.rb]: thừa whitespace: `dependent: :destroy`

File: app/controllers/users_controller.rb (hunk 28 above)
Comment (hungnh103) [inline app/controllers/users_controller.rb]: remove `()`

File: app/controllers/users_controller.rb (hunk 29 above)
Comment (hungnh103) [inline app/controllers/users_controller.rb]: remove `()`

File: app/models/user.rb [hunk 31]
```diff
@@ -35,6 +42,18 @@ def authenticated? attribute, token
     BCrypt::Password.new(digest).is_password?(token)
   end
 
+  def follow other_user
+    following << other_user
+  end
+  
+  def unfollow other_user
+    following.delete(other_user)
```
Comment (huongnguyenmta) [inline app/models/user.rb]: remove ()

File: app/models/user.rb [hunk 32]
```diff
@@ -35,6 +42,18 @@ def authenticated? attribute, token
...
+  end
+
+  def following? other_user
+    following.include?(other_user)
```
Comment (huongnguyenmta) [inline app/models/user.rb]: remove ()

File: app/helpers/relationships_helper.rb [hunk 33]
```diff
@@ -0,0 +1,2 @@
+module RelationshipsHelper
```
Comment (hungnh103) [inline app/helpers/relationships_helper.rb]: remove file

File: app/models/relationship.rb [hunk 34]
```diff
@@ -0,0 +1,6 @@
+class Relationship < ApplicationRecord
+  belongs_to :follower, class_name: "User"
```
Comment (hungnh103) [inline app/models/relationship.rb]: `User.name`

File: app/models/relationship.rb [hunk 35]
```diff
@@ -0,0 +1,6 @@
...
+  belongs_to :followed, class_name: "User"
```
Comment (hungnh103) [inline app/models/relationship.rb]: `User.name`

File: app/views/users/_follow.html.erb [hunk 36]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.build) do |f| %>
```
Comment (huongnguyenmta) [inline app/views/users/_follow.html.erb]: remove ()

File: app/models/user.rb [hunk 37]
```diff
@@ -1,5 +1,12 @@
 class User < ApplicationRecord
   has_many :microposts, dependent: :destroy
+  has_many :active_relationships, class_name: "Relationship",
```
Comment (hungnh103) [inline app/models/user.rb]: `Relationship.name`

File: app/models/user.rb [hunk 38]
```diff
@@ -1,5 +1,12 @@
...
+    foreign_key: "follower_id", dependent: :destroy
+  has_many :passive_relationships, class_name: "Relationship",
```
Comment (hungnh103) [inline app/models/user.rb]: `Relationship.name`

File: app/models/user.rb (hunk 30 above)
Comment (hungnh103) [inline app/models/user.rb]: remove whitespace

File: app/models/user.rb [hunk 39]
```diff
@@ -35,6 +42,18 @@ def authenticated? attribute, token
     BCrypt::Password.new(digest).is_password?(token)
```
Comment (hungnh103) [inline app/models/user.rb]: remove `()`

File: app/models/user.rb (hunk 31 above)
Comment (hungnh103) [inline app/models/user.rb]: remove `()`

File: app/models/user.rb (hunk 32 above)
Comment (hungnh103) [inline app/models/user.rb]: remove `()`

File: app/views/users/_follow_form.html.erb [hunk 40]
```diff
@@ -0,0 +1,9 @@
+<% unless current_user? @user %>
+  <div id="follow_form">
+  <% if current_user.following? @user %>
```
Comment (hungnh103) [inline app/views/users/_follow_form.html.erb:3]: indent

**Pull request merged**